lib/mlir/Dialect/Clift/IR/CliftOps.cpp

```diff
@@ -25,6 +25,28 @@ void mlir::clift::CliftDialect::registerOperations() {
                 /* End of operations list */>();
 }
 
+//===---------------------------- Region types ----------------------------===//
+
+template<typename OpInterface>
+bool clift::impl::verifyRegionContent(Region &R) {
+  if (R.empty())
+    return true;
+
+  if (not R.hasOneBlock())
+    return false;
+
+  const auto Result = R.walk([&](Operation *const NestedOp) -> WalkResult {
+    return mlir::success(mlir::isa<OpInterface>(NestedOp));
+  });
+
+  return not Result.wasInterrupted();
+}
+
+template bool clift::impl::verifyRegionContent<StatementOpInterface>(Region &R);
+
+template bool
+clift::impl::verifyRegionContent<ExpressionOpInterface>(Region &R);
+
```

Comment:
It's very generic defined like this, but I think we could use `verifyRegionContent` to make sure that no operation is both a statement and an expression.

Something like:
```c++
template<bool OnlyStatements>
bool clift::impl::verifyRegionContent(Region &R) {
  if (R.empty())
    return true;

  if (not R.hasOneBlock())
    return false;
  
  using AllowedOpInterface = std::conditional_t<OnlyStatements, StatementOpInterface, ExpressionOpInterface>;
  using ForbiddenOpInterface = std::conditional_t<not OnlyStatements, ExpressionOpInterface, StatementOpInterface;
  const auto Result = R.walk([&](Operation *const NestedOp) -> WalkResult {
    return mlir::success(mlir::isa<AllowedOpInterface>(NestedOp) and not mlir::isa<ForbiddenOpInterface>(NestedOp));
  });

  return not Result.wasInterrupted();
}
```

Then it would be nice to have a way to properly report errors whenever e.g. an operation is both a statement and an expression. Not that I expect anything like that to pass review, but during developement and onboarding of junior devs it can be very helpful.

Reply:
I think we're going to have to split them up into separate functions anyway because the expression region block should be non-empty if present and have a yield as its last operation.

Reply:
As for checking that no operation implements both interfaces, I think this is absolutely not the place to do it. If you want, I can perhaps add a unit test to iterate over all Clift dialect operations and check that none of them implement both.

Reply:
I added those expression region checks I mentioned, but not the one you asked for.

Reply:
Yeah we've discussed about this in person. It's fine like this.